src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets

```diff
@@ -29,6 +29,10 @@ Copyright (c) .NET Foundation. All rights reserved.
   <UsingTask TaskName="GetAssemblyVersion" AssemblyFile="$(MicrosoftNETBuildTasksAssembly)" />
   <UsingTask TaskName="GenerateSatelliteAssemblies" AssemblyFile="$(MicrosoftNETBuildTasksAssembly)" />
 
+  <ItemGroup Condition="'$(BuildWithFrameworkRoslyn)' == 'true' and '$(MicrosoftNetFrameworkCompilersToolsetPackageVersion)' != '' and @(PackageReference->AnyHaveMetadataValue('Identity', 'Microsoft.Net.Compilers.Toolset.Framework')) != 'true'">
+    <PackageReference Include="Microsoft.Net.Compilers.Toolset.Framework" PrivateAssets="all" Version="$(MicrosoftNetFrameworkCompilersToolsetPackageVersion)"/>
```

Comment:
Throwing some out there:

`BuildWithFrameworkRoslyn` -> `BuildWithNetFrameworkHostedCompiler`
`MicrosoftNetFrameworkCompilersToolsetPackageVersion` -> `NetFrameworkHostedCompilersVersion`

Reply:
> and '$(MicrosoftNetFrameworkCompilersToolsetPackageVersion)' != '' and @(PackageReference->AnyHaveMetadataValue('Identity', 'Microsoft.Net.Compilers.Toolset.Framework')) != 'true'"

I feel like this conditions end up masking problems and making it harder for customers to detect issues. Basically it creates two cases where a customer says `<BuildWithFrameworkRoslyn>true</BuildWithFrameworkRoslyn>` but we **silently** do not actually do it. 

Instead can we consider putting a diagnostic here when either of these conditions are met? Essentially very visibly alert the customer that a problem exists.

Reply:
> MicrosoftNetFrameworkCompilersToolsetPackageVersion

How does this property get set on customer installations? Basically when a customer installs the .NET SDK what props file sets this value? Expecting to see a change here that pushes it but don't.

Reply:
> How does this property get set on customer installations? Basically when a customer installs the .NET SDK what props file sets this value? Expecting to see a change here that pushes it but don't.

That's [this](https://github.com/dotnet/installer/blob/4f3e7dba2ff24dff107813d0f85dc38ddb2b3cc9/src/redist/targets/GenerateBundledVersions.targets#L355).

Reply:
> That's [this](https://github.com/dotnet/installer/blob/4f3e7dba2ff24dff107813d0f85dc38ddb2b3cc9/src/redist/targets/GenerateBundledVersions.targets#L355).

Ah okay. So what version of the .NET SDK can I see this value being present? Tried searching some older 8.0 previews and didn't see it. Wanted to just kinda get a feel for where this is. 

> That just means that someone wanted to pin to an older version of the framework compiler, so they added their own PackageReference before this was imported.

Correct but that is essentially the build file in conflict with itself. There are essentially two declarations saying "add this package" and we're silently picking one.

Reply:
8.0.100-preview 3, I think. Does that sound right, @joeloff?

This feels like the most natural way to add an override to me. People are already familiar with adding PackageReferences to various packages and including versions as desired; this just ensures that if they do that, we don't download a wrong version. We could instead recommend that people overwrite the version provided by default in the SDK, but then they'd have to know about our particular version property's name. Do you have a different way you think an override should be introducible?

Reply:
> Do you have a different way you think an override should be introducible?

If people want to manually override they should add the package reference directly and unset the property. The issue with the current design is when we see `<BuildWithFrameworkRoslyn>true</BuildWithFrameworkRoslyn>` we don't actually know what is happening. Customers could be using the framework compiler we introduced or they could be overriding it. Both are valid outcomes. 

That seems incorrect to me. They specified they wanted the framework roslyn, we should give that to them or fail because we can't give it to them. 

If the customer is both setting the property and adding a package reference then they've effectively set conflicting build properties. We should error and tell them to decide which they want.

Reply:
I talked with marcpopMSFT, and we're not convinced that's a problem we have to solve. I agree they could be using the framework compiler _version_ we introduced or overriding the _version_, but they're using the framework compiler either way—only the version shifted—so I still feel the property is valid. It isn't currently being used in this way, but I could easily imagine later using that property to affect other parts of the build.

The big question is how often customers will break themselves by doing something. In this case, it doesn't seem terribly harmful to have overridden the version in addition to asking us to use the default version if they don't override it. That said, not having a version is certainly a problem. I suggested that I'd just remove the check, in which case we try to resolve a PackageReference with no version, which leads to a NuGet warning.

Reply:
> The big question is how often customers will break themselves by doing something

I strongly disagree. I think the question is how much effort does it take from us to stop the customer from hurting themselves in the first place. The effort here is _minimal_. The change to stop a bug from ever happening in the first place is orders of magnitude less than the text we've already written here. 

The change as written is inviting customers to write MSBuild code that does not do as it's stated. The documentation for `$(BuildWithFrameworkRoslyn)` will need to be 

> This might use the Roslyn compiler from the SDK or it might not. 

Why are we fighting making the customer story here significantly simpler?

Reply:
Is the `$(BuildWithFrameworkRoslyn)` used for anything else? If the `<PackageReference/>` exists, it's like an implicit version the user configured and that implies `BuildWithFrameworkRoslyn=true`. However, if the package doesn't exist, but the property is `true`, that indicates the user wants to use the default SDK values.

What if we named the property `BuildWithDefaultFrameworkRoslyn`. If `true` and the user defines a `PackageReference` we error out because there can only be one version, and this happens later in some target, post-evaluation.

I don't know enough to say what the default should be. Assuming it's the SDK copy, then setting the property to true would require an explicit gesture from a user to set it to false and provide a package reference that would also have a specific/latest version attached to it.

The version property might still be an issue. Can we rename it and prefix it with `_` - convention says that's an internal property that users should not override and if they do, they're on their own.

Reply:
FWIW the .NET Framework targeting pack auto package reference does the following: https://github.com/dotnet/sdk/blob/31d20a223b9d3442d0641bb6d54a1d0f56b682a2/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets#L462-L465

Would we good to align with that code to be consistent.

Reply:
> Would we good to align with that code to be consistent. 

Where does the user set a property that explicitly asks for this package to be included? That is my key problem here. We're defining a property whose entire intent is "reference this package" then introducing a case where we silently fail to do that. This is going to make build more confusing, not less.

Reply:
Right, .NET Framework targeting pack auto-references are opt-out instead of opt-in.

> We're defining a property whose entire intent is "reference this package" then introducing a case where we silently fail to do that. This is going to make build more confusing, not less.

Agreed. I suggest to verify if NuGet already emits an error for duplicate package references with `IsImplicitlyDefined` metadata and if not, emit an error ourselves.